### ISSUE TITLE
Use 'frecency' in the inserter?

### ISSUE THREAD

**Issue by noisysocks** [WordPress/gutenberg#5320]
<img width="374" alt="screen shot 2018-03-01 at 10 58 10" src="https://user-images.githubusercontent.com/612155/36819899-770ff72c-1d3f-11e8-9d7a-bc4b2e30d923.png">

Currently, blocks in the _Recent_ tab are ordered from most recently used to least recently used. This works, but fails to take into account how frequently a block is used. It might be worth exploring sorting these blocks by [_frecency_](https://en.wikipedia.org/wiki/Frecency).

I like this definition used by [z](https://github.com/rupa/z):

> Frecency is a portmanteau of ‘recent’ and ‘frequency’. It is a weighted rank  that depends on how often and how recently something occurred. As far as I know, Mozilla came up with the term.
>
> To z, a directory that has low ranking but has been  accessed  recently will  quickly  have  higher rank than a directory accessed frequently a long time ago.

The [sorting function that z uses](https://github.com/rupa/z/blob/ea5ec7834398ee2244de26953b3d1ef785d3f942/z.sh#L138-L145) looks like it might work well for us, as well:

```awk
function frecent(rank, time) {
    # relate frequency and time
    dx = t - time
    if( dx < 3600 ) return rank * 4
    if( dx < 86400 ) return rank * 2
    if( dx < 604800 ) return rank / 2
    return rank / 4
}
```

Thoughts? Is this worth trying? cc. @karmatosed @jasmussen @mtias

**Comment by jasmussen** [comment]
I think this will make @folletto happy :)

Recency is good because it sorts blocks according to those you actually _use_, so blocks you rarely use aren't on top. But at the same time this can mess with spatial memory. If you're used to always having the Image first and the Gallery 2nd, you might stumble if they suddenly switch place. So anything to improve that spatial memory is a win in my book. Does this help that?

**Comment by karmatosed** [comment]
This is so awesome! I am really excited about exploring this and it's something we hadn't considered but would feel magic.

**Comment by danieltj27** [comment]
I really like this idea of having most used blocks. Working out how to approach the per post and per user most used will be interesting though. I can imagine having a most used per page such as text may not be the most used per user such as image for example. Great idea though 👍

**Comment by folletto** [comment]
> I think this will make @folletto happy :)

VERY happy, as it's something I suggested when we did the first design https://github.com/WordPress/gutenberg/issues/888#issuecomment-310009639 :D :D :D 


> But at the same time this can mess with spatial memory. If you're used to always having the Image first and the Gallery 2nd, you might stumble if they suddenly switch place. So anything to improve that spatial memory is a win in my book. Does this help that?

Yes. It's not perfect for spatial memory, but it's a "sweet spot" between purely short-term memory based approaches and spatial memory based approaches. As it makes the list recent but not too quickly changed. :)

> I like this definition used by z:

I had no idea of that elaboration and that algorithm. I think it's a great starting point, and then we can tweak the weightings as we find it's better through testing. :)
